### PR TITLE
CAMEL-14816 Introduce supportLevel and compileTarget in the Catalog

### PR DESCRIPTION
**Pull request by ppalaga**
https://issues.apache.org/jira/browse/CAMEL-14816

### CODE REVIEW

**Comment by ppalaga** [comment]
> CompilationTarget it's something specific to Quarkus. So in my opinion it should be renamed.

I am open to other naming proposals, but I do not think there is much to improve. 

Native compilation is a concept that goes well beyond Quarkus. Yes, Quarkus and GraalVM are nowadays the most prominent tools for the native compilation, but they are not the only conceivable ones.

**Comment by oscerd** [comment]
> > CompilationTarget it's something specific to Quarkus. So in my opinion it should be renamed.
> 
> I am open to other naming proposals, but I do not think there is much to improve.
> 
> Native compilation is a concept that goes well beyond Quarkus. Yes, Quarkus and GraalVM are nowadays the most prominent tools for the native compilation, but they are not the only conceivable ones.

Yes, true, but actually camel as main project doesn't aim to run in native mode, the only subproject focused on that is camel-quarkus: my point is just avoid confusion because this is an information interesting only from a camel-quarkus perspective.

**Comment by oscerd** [comment]
By the way lets wait for other opinions too. Maybe someone has a better idea for naming.

**Comment by ppalaga** [comment]
> camel as main project doesn't aim to run in native mode, the only subproject focused on that is camel-quarkus: my point is just avoid confusion because this is an information interesting only from a camel-quarkus perspective.

I think the picture is bit broader: yes, Camel Quarkus is going to be the only subproject producing `Native` components, but there is Camel K who will leverage (consume) that information. 

Camel still seems to be the best place to host the model, although it won't leverage the info stored in the field.

**Comment by oscerd** [comment]
> > camel as main project doesn't aim to run in native mode, the only subproject focused on that is camel-quarkus: my point is just avoid confusion because this is an information interesting only from a camel-quarkus perspective.
> 
> I think the picture is bit broader: yes, Camel Quarkus is going to be the only subproject producing `Native` components, but there is Camel K who will leverage (consume) that information.
> 
> Camel still seems to be the best place to host the model, although it won't leverage the info stored in the field.

It's just a name to choose basically, maybe compilationTarget is a bit too much generic, but we can always renaming it or not. Lets wait for others opinions.

**Comment by oscerd** [comment]
@davsclaus we can postpone this to 3.3.0, since it has also conflicts.

**Comment by ppalaga** [comment]
The conflicts are only in generated files. They are easy to fix.

**Comment by oscerd** [comment]
Yes, but until we don't have more opinions on this, I'd prefer to wait.

**Comment by ppalaga** [comment]
ba1f0b8 rebased and resolved the conflicts

**Comment by dmvolod** [comment]
I agree @oscerd , maybe use name like runtimeTarget and use this name not only for Quarkus and Quarkus or GraalVM will be just one the option?

**Comment by davsclaus** [comment]
Sorry for the late comment here, but I had to think a bit and also busy with work.

Having details whether a component can be native compiled with graal or not is desirable, the more widespread graal becomes over time. I do think at first this detail belongs in the camel-quarkus catalog as thats the sub project that leads this effort with camel on graal.

For example this PR marks all components as JVM, but some of them would be native compilable out of the box, and others have extensions as part of camel-quarkus that make this possible. So you can then have a conflict here, eg camel-netty marked as JVM here, and then camel-netty marked as native in camel-quarkus.

**Comment by davsclaus** [comment]
In terms of the support level, then that can be benficial.

- experimental
- preview
- stable
- deprecated ??

Are values that Quarkus uses. 

At Camel we have experimental as well but only as annotation for java code. And components can be deprecated. So we dont have a preview and stable marker. But IMHO a problem with stable is that it gives false impression that it "stays like this forever". But Camel is only stable for each LTS release where we keep it stable and do patch releases. Between LTS releases we have influence to move Camel forward and change components. For example often by 3rd party dependency upgrades that can cause the component options to change etc. So I think the "stable" would be over reaching.

So instead we need a different word (default). 

This PR just adds a bulk of settings for everything but dont offer a way of configuring this individually per component (how to mark something as preview etc).

I therefore think this needs more debate on Camel dev mailing list and that this PR can be parked until then.

**Comment by davsclaus** [comment]
tl;dr

I think something like this benefits Camel. But needs some more thought and we should do this on Camel dev mailing list. 

Also another idea could be to mark JVM levels, eg we had in the past a few components that didnt work on Java 8 but only on 7 etc. So we can have this again with 8 vs 11, and in the future with 11 vs 14 etc. So should we have a "value" for this in the catalog?

**Comment by davsclaus** [comment]
This is being discussed at the Camel dev mailing list
https://camel.465427.n5.nabble.com/DISCUSS-Camel-components-Information-about-stability-and-native-compilation-td5863063.html

**Comment by davsclaus** [comment]
Peter would it have been possible to have separated this PR into multiple commits, one for the tooling change, and another for the source code updates in json files, as the latter can be regen safely and as this is one commit then we cant easily grab the tooling changes as we get massive merge conflicts.

**Comment by davsclaus** [comment]
Okay we have now the supportLevel in the master branch. Lets give some more thought to the jvm level option at the dev mailing list.

Closing this PR for now.

**Comment by ppalaga** [comment]
Thanks Claus for taking over. I was busy with other tasks yesterday.
